File: solutions/873/main_tle.py
from typing import * 

# TLE 
from collections import defaultdict 
class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        n = len(arr)
        dp = [[0] * n for _ in range(n)]
        if arr[2] == arr[0] + arr[1]: 
            dp[2][1] = 3
        for i in range(3, n): 
            for j in range(i):
                for k in range(j): 
                    if arr[j] + arr[k] == arr[i]:
                        inc =  dp[j][k]+1 if dp[j][k] != 0 else 3
                        dp[i][j] = max(dp[i][j], inc)
        return max(map(max, dp)) 

if __name__ == "__main__": 
    s = Solution()

    # arr = [1,2,3,4,5,6,7,8]
    # assert s.lenLongestFibSubseq(arr) == 5 

    # arr = [1,3,7,11,12,14,18]
    # assert s.lenLongestFibSubseq(arr) == 3 

    arr = [1,3,4,7,10,11,12,18,23,35]
    assert s.lenLongestFibSubseq(arr) == 6  

File: solutions/873/main_tle.rs
// TLE
impl Solution {
    pub fn len_longest_fib_subseq(arr: Vec<i32>) -> i32 {
        let n = arr.len(); 
        let mut dp = vec![vec![0; n]; n]; 
        let mut result = 0; 
        if arr[2] == arr[0] + arr[1] {
            dp[2][1] = 3; 
            result = 3; 
        }
        for i in 3..n {
            for j in 0..i {
                for k in 0..j {
                    if arr[k] + arr[j] == arr[i] {
                        if dp[j][k] != 0 {
                            dp[i][j] = dp[i][j].max(dp[j][k]+1);
                        }
                        else {
                            dp[i][j] = 3; 
                        }
                        result = result.max(dp[i][j]); 
                    }
                }
            }
        }
        return result; 
    }
}

struct Solution {}

fn main() {
    assert_eq!(Solution::len_longest_fib_subseq(vec![1,2,3,4,5,6,7,8]), 5); 
}